### ISSUE TITLE
Use CARGO_TARGET_TMPDIR if set

### ISSUE THREAD

**Issue by dimo414** [flub/testdir#2]
Cargo provides a temp directory for integration tests via [`CARGO_TARGET_TMPDIR`](https://doc.rust-lang.org/cargo/reference/environment-variables.html). It's unfortunately not currently set for unit tests but it might still be nice to use it when it's available (I am only using testdir with integration tests, and I imagine that's somewhat typical). https://github.com/rust-lang/cargo/pull/9375 has some context behind the motivation for this variable and why it's only set for integration tests.

FWIW I went looking for this setting because I was a little surprised testdir creates all its directories right in the `target/` root; it ends up creating nine entries that seem like they could be nested under a common parent directory to keep the root dir cleaner. So in addition to / instead of using `CARGO_TARGET_TMPDIR` it might be nice to put all these paths under a single common directory such as `target/testdir/dir-*`.

**Comment by flub** [comment]
> So in addition to / instead of using CARGO_TARGET_TMPDIR it might be nice to put all these paths under a single common directory such as target/testdir/dir-*.

Mind filing a separate issue for this?

**Comment by flub** [comment]
Thanks for the reference.  Doing this I guess means that testdir would behave differently when running integration tests as when running unittests because cargo will always set this directory for integration tests?  So now you'd have two paths where you have to go look depending on which tests you happened to run.  I'm not entirely sure how much I like this.

**Comment by dimo414** [comment]
Yeah I agree it's too bad the var isn't consistently set (it sounds like others would like it too), but it might still make sense to respect it if/when it is set, even if that's not always the case.

**Comment by flub** [comment]
I'm re-reading this issue and motivation, and I'm inclined to not use this variable.  I think the CARGO_TARGET_TMPDIR environment variable is an alternative, arguably more primitive, way to find a persistent temporary directory.  It has weaker guarantees around it, e.g. about cleaning up the directory and managing subdirectories.  One motivation for CARGO_TARGET_TMPDIR is that it is hard to find the target directory, which is something testdir already figures out.

I think it would not be unlikely that using CARGO_TARGET_TMPDIR would run into conflicts of other code mucking about in that directory, causing more confusing situations.  So I think it is preferably to keep testdir's directory separate.
